paasta_tools/autoscaling/autoscaling_cluster_lib.py

```diff
@@ -364,69 +364,97 @@ def scale_aws_spot_fleet_request(resource, current_capacity, target_capacity, po
-            # we giveup waiting for the instance to drain and mark it for termination anyway
-            start = int(time.time() + drain_timeout) * 1000000000  # nanoseconds
-            # Set the duration to an hour, this is fairly arbitrary as mesos doesn't actually
-            # do anything at the end of the maintenance window.
-            duration = 600 * 1000000000  # nanoseconds
-            log.info("Draining {0}".format(slave_to_kill['pid']))
-            if not dry_run:
-                try:
-                    drain_host_string = "{0}|{1}".format(slave_to_kill['hostname'], slave_to_kill['ip'])
-                    drain([drain_host_string], start, duration)
-                except HTTPError as e:
-                    log.error("Failed to start drain "
-                              "on {0}: {1}\n Trying next host".format(slave_to_kill['hostname'], e))
-                    continue
-            log.info("Decreasing spot fleet capacity from {0} to: {1}".format(current_capacity, new_capacity))
-            if not set_spot_fleet_request_capacity(sfr_id, new_capacity, dry_run, region=resource['region']):
-                log.error("Couldn't update spot fleet, stopping autoscaler")
-                log.info("Undraining {0}".format(slave_to_kill['pid']))
-                if not dry_run:
-                    undrain([drain_host_string])
-                break
-            log.info("Waiting for instance to drain before we terminate")
-            try:
-                wait_and_terminate(slave_to_kill, drain_timeout, dry_run, region=resource['region'])
-            except ClientError as e:
-                log.error("Failure when terminating: {0}: {1}".format(slave['pid'], e))
-                log.error("Setting spot fleet capacity back to {0}".format(current_capacity))
-                if not set_spot_fleet_request_capacity(sfr_id, current_capacity, dry_run, region=resource['region']):
-                    log.error("Couldn't update spot fleet, stopping autoscaler")
-                    break
-                continue
-            finally:
-                log.info("Undraining {0}".format(slave_to_kill['pid']))
-                if not dry_run:
-                    undrain([drain_host_string])
-            current_capacity = new_capacity
-            mesos_state = get_mesos_state_summary_from_leader()
-            filtered_slaves = get_mesos_task_count_by_slave(mesos_state, slaves_list=filtered_sorted_slaves)
+        downscale_spot_fleet_request(resource=resource,
+                                     filtered_slaves=filtered_slaves,
+                                     current_capacity=current_capacity,
+                                     target_capacity=target_capacity,
+                                     pool_settings=pool_settings,
+                                     dry_run=dry_run)
+
+
+def gracefully_terminate_slave(resource, slave_to_kill, pool_settings, current_capacity, new_capacity, dry_run):
```

Comment:
Maybe on the next pass this function can be split up into two pieces, one to gracefully drain the slave, and the other to terminate + scale the sfr?

The first half of the code deals with draining and the slave_to_kill, the second part of this function deals with new_cap, current_cap, etc.